zebrad/src/components/mempool/downloads.rs

```diff
@@ -0,0 +1,284 @@
+/// Therefore, this attack can be carried out by a single malicious node.
+const MAX_INBOUND_CONCURRENCY: usize = 10;
+
+/// The action taken in response to a peer's gossiped transaction hash.
+pub enum DownloadAction {
+    /// The transaction hash was successfully queued for download and verification.
+    AddedToQueue,
+
+    /// The transaction hash is already queued, so this request was ignored.
+    ///
+    /// Another peer has already gossiped the same hash to us.
+    AlreadyQueued,
+
+    /// The queue is at capacity, so this request was ignored.
+    ///
+    /// The sync service should discover this transaction later, when we are closer
+    /// to the tip. The queue's capacity is [`MAX_INBOUND_CONCURRENCY`].
+    FullQueue,
+}
+
+/// Represents a [`Stream`] of download and verification tasks.
+#[pin_project]
+#[derive(Debug)]
+pub struct Downloads<ZN, ZV, ZS>
+where
+    ZN: Service<zn::Request, Response = zn::Response, Error = BoxError> + Send + 'static,
+    ZN::Future: Send,
+    ZV: Service<tx::Request, Response = tx::Response, Error = BoxError> + Send + Clone + 'static,
+    ZV::Future: Send,
+    ZS: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + Clone + 'static,
+    ZS::Future: Send,
+{
+    // Services
+    /// A service that forwards requests to connected peers, and returns their
+    /// responses.
+    network: ZN,
+
+    /// A service that verifies downloaded transactions.
+    verifier: ZV,
+
+    /// A service that manages cached blockchain state.
+    state: ZS,
+
+    // Internal downloads state
+    /// A list of pending transaction download and verify tasks.
+    #[pin]
+    pending: FuturesUnordered<JoinHandle<Result<UnminedTxId, (BoxError, UnminedTxId)>>>,
+
+    /// A list of channels that can be used to cancel pending transaction download and
+    /// verify tasks.
+    cancel_handles: HashMap<UnminedTxId, oneshot::Sender<()>>,
+}
+
+impl<ZN, ZV, ZS> Stream for Downloads<ZN, ZV, ZS>
+where
+    ZN: Service<zn::Request, Response = zn::Response, Error = BoxError> + Send + Clone + 'static,
+    ZN::Future: Send,
+    ZV: Service<tx::Request, Response = tx::Response, Error = BoxError> + Send + Clone + 'static,
+    ZV::Future: Send,
+    ZS: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + Clone + 'static,
+    ZS::Future: Send,
+{
+    type Item = Result<UnminedTxId, BoxError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        // CORRECTNESS
+        //
+        // The current task must be scheduled for wakeup every time we return
+        // `Poll::Pending`.
+        //
+        // If no download and verify tasks have exited since the last poll, this
+        // task is scheduled for wakeup when the next task becomes ready.
+        //
+        // TODO:
+        // This would be cleaner with poll_map #63514, but that's nightly only.
```

Comment:
```suggestion
        // TODO: this would be cleaner with poll_map (#2693)
```

`poll_map` was stabilised in January 2021:
https://github.com/rust-lang/rust/issues/63514
https://doc.rust-lang.org/nightly/std/task/enum.Poll.html#method.map_ok-1

I opened ticket #2693.

Can you update the corresponding comments in the sync and inbound block downloaders?

Reply:
Done in